### PR TITLE
Nano integration V2

### PR DESCRIPTION
**Pull request by koczadly**
Second revision of the integration for [Nano](https://nano.org) cryptocurrency. Modification of #488 

Reimplemented without any additional dependencies. Note that I did add `com.squareup.okio` and `org.java-websocket` to the extra `build.gradle`, but these dependencies have already been approved and accepted in the `dependencyChecksums.txt` file (using the same version), presumably used in another one of your modules.

If you have any recommendations or proposed changes please let me know so I can rectify them ASAP. Cheers

### CODE REVIEW

**Comment by koczadly** [comment]
@generalbytes Any updates on the approval of this integration? Is there anything that needs amending before you can accept it? Sorry to push but there's at least one operator/merchant waiting on the approval and deployment of this integration.

Cheers

**Comment by koczadly** [comment]
@generalbytes Merge conflicts have been resolved, please advise of any other changes you require to merge this into the repo.

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java
Comment (pvyhnal-generalbytes) [inline server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java:79]: would this be OK if this extension is used by multiple organizations (operators) with multiple different rpc wallets on the same server? Why is this needed?

**Review by pvyhnal-generalbytes — CHANGES_REQUESTED** [review]
See my comment about the shared RPC wallet. Would it be able to make it work without it? Otherwise it looks good

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java
Comment (koczadly) [reply server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java:79]: Hey mate, thanks for reviewing the PR!

A 'global' `NanoRpcClient` instance is used as Nano relies on some additional cryptographic functions which were rejected in a former PR as multiple libraries are required — to get around this, I changed it so the node processes all the cryptographic work instead of being done locally on the CAS server. This does however mean that to use basic features like the paper wallet generator, a `nano_node` wallet is still required and needs to be configured by the operator. Not a very pretty way of doing things I'll admit, but it's the only solution I could think up at the time.

After reading some of the documentation, I was under the impression that each organisation would use their own independent CAS server (and Nano node) rather than being a shared system — if this isn't the case or there are other security considerations that need to be make, please do let me know and I'll review what the options are.

Comment (pvyhnal-generalbytes) [reply server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java:79]: The operators can choose to host their own CAS server (and in that case they probably would have just one Nano node) but if we merge this PR into our release it will also be available for all the operators using our "cloud" hosted CAS server that is shared between different operators. In that case everything has to be completely separate for each of them. 

Would a more basic implementation without the global RPC node work? For example without the support for paper wallets or just a basic address validation? Or is paper wallet support etc really required? What other features depend on it?

Another option would be to keep the extension private (not merged in our product) and you (the operator?) would install it only on their own server.

Comment (koczadly) [reply server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java:79]: Ah I see, I didn't realise that was an option. With the cloud CAS service, would the Nano node be ran by yourselves and shared, or would each ATM operator configure and run their own node?

As for support of paper wallets, I'm honestly not sure if it's a requirement. I was under the impression that paper wallets are an essential feature to purchase the crypto, but I honestly haven't seen one of your ATMs in use so it's a bit hard to judge what features may be required.

Could you explain further on how to install the extension privately? That could work for the time being, as we have someone wanting to get this up and running ASAP. I always figured the CAS system would require signing of the JARs as a security protocol, so never thought of it as an option.

**Comment by pvyhnal-generalbytes** [comment]
On cloud CAS each operator configures their own wallets / nodes.

The standard procedure for Cash-to-fiat is that the buyer shows a QR code
with their destination address to the ATM's camera, insert cash, and the
server sends the coins to the scanned address.
Lot of the coins we support don't have a paper wallet option but that
depends on your requirements

By installing the extension privately I meant that you can build the JAR
yourself (from this branch) and just put it in the `extensions` folder in
your CAS server installation if you don't use our cloud

If you need this on our cloud, I'd recommend just a more basic
implementation without the shared global node.

**Comment by koczadly** [comment]
@pvyhnal-generalbytes If multiple operators can share a single CAS server, then using a global RPC node for certain activities would be a security concern.

I have removed those uses and features from the integration and pushed the commit, it should not be ready for review and merging. If you have any other questions or concerns about it just let me know and I can make further changes if required. Thanks for helping us out!

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java
Comment (pvyhnal-generalbytes) [inline server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java:113]: Please don't add the rate sources that are already created in other extensions. 
If you add NANO to coinmarketcap rate source for example and `new CoinmarketcapRateSource` is called in another extension it will work fine.

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/test/TestNodeWallet.java
Comment (pvyhnal-generalbytes) [inline server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/test/TestNodeWallet.java:17]: could you please remove these classes or move them to `/src/test/` as `@Ignore`-d tests for example?

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java
Comment (koczadly) [reply server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java:113]: Wouldn't that mean our extension is coupled to the Dash implementation for instance? So for example, if `DashExtension` specifies `CoinmarketcapRateSource` as of now, and removes it at a later date, that will affect the behaviour of the `NanoExtension` (and any other extension that relies on that source). Almost all extensions seem to re-specify the rate sources (at least 15 extensions).

That being said, I'll remove it for the time being if that's what you'd prefer.

Comment (pvyhnal-generalbytes) [reply server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/nano/NanoExtension.java]: Thanks. You're right coinmarketcap is re-specified where it's not necessary in other extensions. But we will not remove it in future, all the extensions in this repository are distributed together as part of the "server_extensions_extra" extensions JAR file.

**Review by pvyhnal-generalbytes — APPROVED** [review]

**Comment by koczadly** [comment]
@pvyhnal-generalbytes Awesome, cheers mate! Do you have any ideas on how long this will take to be available to ATM operators?

**Comment by pvyhnal-generalbytes** [comment]
> @pvyhnal-generalbytes Awesome, cheers mate! Do you have any ideas on how long this will take to be available to ATM operators?

It may take around 3 weeks to finish testing and release, you can follow our releases here: https://www.generalbytes.com/en/support/changelog

If the operator run their own server you can build extensions .jar from this repository and the operator can try to test it themselves earlier

**Review by filipocelka — APPROVED** [review]

**Pull request merged**
